### ISSUE TITLE
Extension: Scalar block layout

### ISSUE THREAD

**Issue by magcius** [gpuweb/gpuweb#4040]
It would be nice to have support for scalar block layout. It's pretty convenient and natural to have packed alignment for structures, and it's a frequent stumbling block for new learners, especially the awkward rules around vec3.

**Comment by greggman** [comment]
I think it's far more likely reflection will be added. The alignment requirements are their for a reason. Also there are already libraries that handle this for you like [this one](https://github.com/greggman/webgpu-utils) and sites that will show you layout like [this one](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d00000100bf06000000000000003d8888623728a306fc320e1a9ba574a7281bf711b8e0fe2ad3ce912bab0e0e45d434d61bb7773cbc401503b1a5422bc9bea039d06b8d7965bfdc6e368ea7d5dafe3438cbb62b089a4d9cec6dc7979747e1b214b3e964bcbcbf7488c72aa713fe7c5d73817e3b3f445d649816a64ac8072c28ba8c3dce046ddbe50ca41aeaa4c22ead3b60c0acb96d367309eaec44952e3088be8405bf4fae4fc8cca8764a61cee989b43b51c4fcd21564ea233ad813d4a383f2d702e9cc9f4da3f8a794715769609bd818144076ed26f05db15962b41006bf38ecbf895e7b023a685147404e2df4271e03bafe88abf306b632c6d273f35a6ce5834950df6a91a3c65d18d097bb0b969d43cd221640e4d91849aadd8c49cf8e9a50a3c54a18d3abd4d1eab3b67735c03defbbd559a482dd5b2900e88233e3d4770a27cd308999e492f60a10802bc15f63de053c189868490b2e480a402fab162e2d699368d63173ecdbfb97a1e6439e89cd0e537ca5e26c9b3bef0be4044e8c2c3125ea6ee7af49ccd727d8ae630486fe523b394b3ca36f6a1aee583ac1f966fe526030a75e34d66d14a36c33a9be1d7fca56ba31f2913c9e3a86e09e9f960d938951253a2695c1cc49e8f3a177b9aa11b2ebd7676123c5786a3ef751e0e9287994ff1d261df33bc90e2fd49262fdc078fc28f3bf4826f1b77d14a4cd85a160e217b5b6ff177338abe3a384350cf06e7a45a01ded9a8c883eb170c5e9887cf843ab7c1c05e7a0f03f0d89c3a2f511bed94c73b34a5b8e0f30de188af06839e2f80e7d5c7c3dcd2cf6189578939e2652a4604fb5a6142a9e2746ce0fa7d0f6125755a75616b8ea0c353e74e2594e630004dd763574a8f6bcf672b365fee19362)

**Comment by kainino0x** [comment]
IIRC scalar layout is something everyone wanted, it's just a matter of hardware reach. I don't remember the reach analysis results, but it's also been a while since we did one.

It's also going to look different from how we initially imagined when we had required offsets/strides on everything. I'm not sure if we're still pursuing the idea of having only one layout algorithm but multiple sets of layout-validity rules, or if we would go for multiple layout algorithms.
The difference being in essence:
- `var<storage> data: @stride(12) array<vec3f>;` vs
- `@layout(scalar) var<storage> data: array<vec3f>;`

**Comment by magcius** [comment]
I'm guessing most of the hold-up is basically old Adreno/Mali, but it would be nice to run the query.

I think the important case is if you have a strided struct like

```rust
@layout(scalar) struct Vertex {
   vec3f position,
   vec3f normal,
   vec2f uv,
};

var<storage> data: array<Vertex>;
```

The natural stride is 3+3+2 = 8 floats / 32 bytes, rather than the 64 bytes it is right now. Where we put the decoration, and how it's spelled, I don't care, but `@stride(32)` isn't quite enough because it doesn't spell the packing.

**Comment by ben-clayton** [comment]
A few drive by comments:

> the struct has an intrinsic layout which you can override at a low level

The spec supports this already. We just need an `enable` to relax layout validation. As mentioned in https://github.com/gpuweb/gpuweb/issues/3756, you could use `align(1)` for all members, and it'll pack to the byte level.

> and/or the struct has an intrinsic layout which you can override at higher levels

While I appreciate this is nice that you don't have to spray `@align` everywhere, this runs counter to one of the nice things about WGSL - there's *one rule* for all buffer layouts. The user doesn't have to remember all the subtle differences of multiple layouts. I do appreciate that scalar layout is far simpler than the std layouts, but I'm guessing there will still be layout complexity for smaller data types (e.g. f16 followed by f32)

> vs the struct does not have an intrinsic layout and the layout is defined by its usage

Here be dragons. This means you end up with an entirely new type for your variable. This introduces all sorts of questions about assignability between different layouts, value equality, etc. I'd oppose this option.

**Comment by kainino0x** [comment]
Got it, I wasn't sure where the spec ended up landing. Agreed on all points.

**Comment by JMS55** [comment]
I have a real world use case of storing a large array of vertices accessed in a compute shader https://github.com/JMS55/bevy/blob/solari/crates/bevy_solari/src/scene/scene_types.wgsl#L16-L39.

I'm not sure I understand the previous discussion - are you saying if I added `@align(4)` to one of the fields I wouldn't have to do the manual unpacking?

**Comment by ben-clayton** [comment]
> I'm not sure I understand the previous discussion - are you saying if I added `@align(4)` to one of the fields I wouldn't have to do the manual unpacking?

Sorry for the confusion.

WGSL has (currently) one rule for buffer layouts. Without a `@align` attribute, the member will use the type's default alignment [as described here](https://www.w3.org/TR/WGSL/#structure-member-layout). With an `@align` attribute, you're explicitly telling the compiler what alignment the member has, overriding its type-inferred default. `@align(4)` will cause the member to align to the next whole scalar - **however** once layout has been done, the buffer is then [validated by its address mode usage(s)](https://www.w3.org/TR/WGSL/#address-space-layout-constraints), which enforces the uniform and storage buffer layout constraints.

So, my previous comment was stating that we already have the language to describe a packed structure layout, but what we're missing is an `enable` directive to relax the layout validation (and enable whatever backend features to make it legal for the underlying API).

I hope this helps.

**Comment by JMS55** [comment]
That clears it up for me, thank you!

**Comment by dneto0** [comment]
For concreteness, here are some cases that I used to get my head straight on this.  Where "today" is after #4978 lands.
It shows that scalar block layout is still easily spec'd if/when we do decide to support it.

```
// SBL = scalar block layout.  Changes RequiredAlignOf(vec4<T>) to sizeof(T)
// 'align' after a struct is alignment of the struct type.

struct Natural     { a: u32,            b: vec4u };  // offsets 0 16, align 16  // should pass
struct Force4      { a: u32, @align(4)  b: vec4u };  // offsets 0 4,  align 4.  // should fail today, pass with SBL
struct Force64     { a: u32, @align(64) b: vec4u };  // offsets 0 64, align 64  // should pass

// nesting
struct Natural_Natural  { nestA: u32,             nestB: Natural } // offsets 0 16,  align 16  // should pass
struct Force4__Natural  { nestA: u32, @align(4)   nestB: Natural } // offsets 0 4,   align 4   // should fail today, pass with SBL
struct Force64_Natural  { nestA: u32, @align(64)  nestB: Natural } // offsets 0 64,  align 64  // should pass

struct Natural_Force4   { nestA: u32,             nestB: Force4  } // offsets 0, 4,  align 4   // should fail today, pass with SBL
struct Force4__Force4   { nestA: u32, @align(4)   nestB: Force4  } // offsets 0, 4,  align 4   // should fail today, pass with SBL
struct Force64_Force4   { nestA: u32, @align(64)  nestB: Force4  } // offsets 0, 64, align 64  // should pass

struct Natural_Force64  { nestA: u32,             nestB: Force64 } // offsets 0, 64, align 64  // should pass
struct Force4__Force64  { nestA: u32, @align(4)   nestB: Force64 } // offsets 0, 4,  align 4   // fail: 4 does not divide align(Force64)
struct Force64_Force64  { nestA: u32, @align(64)  nestB: Force64 } // offsets 0, 64, align 64  // should pass

```

**Comment by dneto0** [comment]
Now that #4978 has landed, the spec change for this is:

- first determine if this is to be available everywhere. See emulation considerations at https://github.com/gpuweb/gpuweb/issues/4973#issuecomment-2491471946 
- change `RequiredAlignOf(vecN<T>)` to `RequiredAlignOf(T)`

**Comment by magcius** [comment]
In either this or #4973, has anybody run a query to figure out what layouts are available on devices shipping WebGPU?
